10.GK/NguyenNgocChung/readme.md

```diff
@@ -0,0 +1,285 @@
+
+    def test_update_data_api(self):
+        data = {
+            "id": "100",
+            "name": "chungngoc",
+            "username": "chungnn",
+            "birth": "2002",
+            "sex": "Male",
+            "university": "PTIT",
+            "major": "dtvt"
+        }
+        response = requests.post("http://localhost:80/api/update/", json=data)
+        responseJson = response.json()
+
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(responseJson["id"], "100")
+
+    def test_read_data_api(self):
+        response = requests.get("http://localhost:80/api/read/15")
+
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(response.json()[0]['id'], '15')
+
+    def test_delete_data_api(self):
+        response = requests.get("http://localhost:80/api/delete/0")
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(response.text, "Delete successfully")
+
+if __name__ == '__main__':
+    unittest.main()
+```
+- Kết quả test API :
+![unit_test](Images/unit_test.jpg)
+
+## **II. Triển khai web application sử dụng các DevOps tools & practices**
+
+### **1. Containerization**
+
+- Docker file cho api:
+```yaml
+FROM python:3.9-alpine
+
+WORKDIR /app
+COPY . /app
+
+RUN pip install -r requirements.txt
+
+CMD ["python","-u", "app.py"]
+```
+- Docker file cho database sử dụng Mongo và đẩy dữ liệu ban đầu vào database:
+```yaml
+FROM mongo:5.0
+
+COPY ./init-db.js /docker-entrypoint-initdb.d/init-db.js
+
+EXPOSE 27017
+
+```
+- Docker file cho web:
+```yaml
+FROM nginx:1.22.0-alpine
+
+COPY ./nginx.conf /etc/nginx/conf.d/default.conf
+
+COPY ./frontend/index.html /usr/share/nginx/html
+COPY ./frontend/update.html /usr/share/nginx/html
+COPY ./frontend/create.html /usr/share/nginx/html
+
+EXPOSE 80
+```
+Docker history Nginx:
+![his_nginx](Images/his_nginx.jpg)
+
+Docker history database:
+![his_db](Images/his_db.jpg)
+
+Docker history API:
+![his_api](Images/his_api.jpg)
+
+
+Dung lượng sau khi build images:
+
+![images_dc](Images/images_dc.jpg)
+
+
+### **2. Continuous Integration**
+
+Sử dụng github action để cấu hình CI. Khi có một push commit lên một branch hoặc PR vào main sẽ tự động chạy unit test.
+
+`.github/workflows/CI.yaml`: 
+
+```yaml
+name: CI
+
+on:
+  push:
+  pull_request:
+    branches:
+      - main
+
+jobs:
+  build:
+    runs-on: ubuntu-latest  
+    steps:
+      - name: Check out the repo
+        uses: actions/checkout@v3
+
+      - name: Run container
+        run: |
+          sudo docker compose build
+          sudo docker compose up -d
+      
+      - name: check container
+        run: sudo docker ps 
+
+
+      - name: Test api
+        run:  python test_api.py
+
+```
+Out put log của luồng CI:
+![CI](Images/CI.jpg)
+
+### ** 3. Continuous Delivery **
+
+Bài giữa kì này em sử dụng ansible để deploy trang web tại 1 node là localhost. Sử dụng roles common để setup docker lên máy.

```

Comment:
Nếu em dùng ansible mà chỉ deploy ở localhost thì coi như mất hết tính chất của 1 IT automation tool rồi, thử thêm deploy trên các node khác nhé